File: hello-typescript/src/logger.ts
/**
 * Logging utility with support for different log levels
 */

let verboseEnabled = false;

export function init(verbose: boolean): void {
  verboseEnabled = verbose;
}

function formatTimestamp(): string {
  const now = new Date();
  return now.toISOString().replace("T", " ").slice(0, 23);
}

export function info(message: string): void {
  console.log(`[${formatTimestamp()}] [info] ${message}`);
}

export function error(message: string): void {
  console.error(`[${formatTimestamp()}] [error] ${message}`);
}

export function warn(message: string): void {
  console.warn(`[${formatTimestamp()}] [warn] ${message}`);
}

export function debug(message: string): void {
  if (verboseEnabled) {
    console.log(`[${formatTimestamp()}] [debug] ${message}`);
  }
}

export function phase(message: string): void {
  console.log();
  console.log(`[${formatTimestamp()}] [info] === ${message} ===`);
}

File: hello-rust/src/logger.rs
use chrono::Local;

static mut VERBOSE: bool = false;

pub fn init(verbose: bool) {
    unsafe {
        VERBOSE = verbose;
    }
}

fn is_verbose() -> bool {
    unsafe { VERBOSE }
}

fn format_timestamp() -> String {
    Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string()
}

pub fn log_debug(message: &str) {
    if is_verbose() {
        println!("[{}] [debug] {}", format_timestamp(), message);
    }
}

pub fn log_info(message: &str) {
    println!("[{}] [info] {}", format_timestamp(), message);
}

pub fn log_warn(message: &str) {
    println!("[{}] [warn] {}", format_timestamp(), message);
}

pub fn log_error(message: &str) {
    eprintln!("[{}] [error] {}", format_timestamp(), message);
}

pub fn log_phase(phase: &str) {
    println!();
    println!("[{}] [info] === {} ===", format_timestamp(), phase);
}
